### ISSUE TITLE
[FSDP] Add unit tests to ensure backward prefetch actually prefetches

### ISSUE THREAD

**Issue by awgu** [pytorch/pytorch#106030]
Per title.

cc @zhaojuanmao @mrshenli @rohan-varma

**Comment by hanzhi713** [comment]
This would be nice, because backward prefetch seems broken in the latest nightly (20230726), even without gradient checkpointing

![image](https://github.com/pytorch/pytorch/assets/21034047/f1fdf1d0-53ad-418a-9dd0-8502b5db1cb3)

**Comment by awgu** [comment]
@hanzhi713 Yes, I landed the revert yesterday morning for the offending PR, so backward prefetch should be back in the next nightly. Sorry about that!

**Comment by hanzhi713** [comment]
It's alright. I should've pinned the version of my previous working nightly.
